drgn/internal/sudohelper.py

```diff
@@ -0,0 +1,88 @@
+# Copyright (c) Stephen Brennan <stephen@brennan.io>
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+"""Helper for opening a file as root and transmitting it via unix socket"""
+import array
+import os
+from pathlib import Path
+import pickle
+import socket
+import subprocess
+import sys
+import tempfile
+from typing import Union
+
+
+def recv_fd(sock: socket.socket) -> int:
+    """Receive a file descripter on the socket"""
+    # Borrows heavily from the Python docs for sock.recvmsg()
+    fds = array.array("i")
+    msg, ancdata, flags, addr = sock.recvmsg(4096, socket.CMSG_SPACE(fds.itemsize))
+    if msg == b"success" and ancdata:
+        level, typ, data = ancdata[0]
+        assert level == socket.SOL_SOCKET
+        assert typ == socket.SCM_RIGHTS
+        data = data[: fds.itemsize]
+        fds.frombytes(data)
+        return fds[0]
+    else:
+        try:
+            exc = pickle.loads(msg)
+        except Exception:
+            exc = Exception(msg.decode())
+        raise exc
```

Comment:
Instead of asserting the level and type, let's scan through the list and look for one that matches what we want, and fail if we don't find one. Also, we might as well blow up if we can't unpickle the message, because `msg.decode()` likely won't work either.

```suggestion
    fds = array.array("i")
    msg, ancdata, flags, addr = sock.recvmsg(4096, socket.CMSG_SPACE(fds.itemsize))
    for level, typ, data in ancdata:
        if level == socket.SOL_SOCKET and typ == socket.SCM_RIGHTS:
            data = data[: fds.itemsize]
            fds.frombytes(data)
            return fds[0]
    raise pickle.loads(msg)
```

At this point, this is small enough that I'd prefer to shove it in `open_via_sudo()` rather than its own function.